### PR TITLE
Prevent special whitespaces in the names of entities

### PR DESCRIPTION
**Pull request by avishniakov**
## Describe changes
I added masking using the banned set of symbols (only special characters `\t\n\r\v\f` are considered banned with this PR) for tags, models, artifacts, stacks, and stack components.
This impacts only the creation of new entities - old ones with invalid names will remain to work, as before, but new ones cannot be created using banned symbols.

Original motivation: symbols like `\n` would be allowed in API calls, but it breaks the Dashboard subsequently.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [x] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

## Summary by CodeRabbit

- **New Features**
  - Enhanced name validation across various components and requests to ensure names contain only allowed characters.
  
- **Bug Fixes**
  - Implemented additional name validation logic in artifact, component, model, model version, stack, and tag requests to prevent invalid name entries.

- **Tests**
  - Added new tests to verify that invalid names are correctly handled and trigger exceptions in various system components.

### CODE REVIEW

**Comment by strickvl** [comment]
@coderabbitai review

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 1]
```diff
@@ -2634,6 +2651,31 @@ def test_register_stack_fails_when_stack_exists():
                     )
 
 
+def test_register_stack_fails_with_invalid_name():
+    """Tests registering stack fails with invalid name."""
+    client = Client()
+    store = client.zen_store
```
Comment (strickvl) [inline tests/integration/functional/zen_stores/test_zen_store.py:2657]:
```suggestion
    store = Client().zen_store
```

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 2]
```diff
@@ -2497,6 +2497,23 @@ def test_count_stack_components():
         assert store.count_stack_components(filter_model) == count_before + 1
 
 
+def test_stack_component_create_fails_with_invalid_name():
+    """Tests that creating a stack component with an invalid name fails."""
+    client = Client()
+    store = client.zen_store
```
Comment (strickvl) [inline tests/integration/functional/zen_stores/test_zen_store.py:2503]:
```suggestion
    store = Client().zen_store
```

**Review by strickvl — COMMENTED** [review]
Just nits. Otherwise looks good to me. I look to @bcdurak for the Pydantic perspective, however.

File: tests/integration/functional/zen_stores/test_zen_store.py (hunk 1 above)
Comment (avishniakov) [reply tests/integration/functional/zen_stores/test_zen_store.py:2657]: USed below: `workspace=client.active_workspace.id`

File: tests/integration/functional/zen_stores/test_zen_store.py (hunk 2 above)
Comment (avishniakov) [reply tests/integration/functional/zen_stores/test_zen_store.py:2503]: Used below: `client.active_workspace.id`

File: tests/integration/functional/zen_stores/test_zen_store.py [hunk 3]
```diff
@@ -4051,6 +4100,19 @@ def test_list_by_tag(self, clean_client: "Client"):
             )
             assert len(ms) == 0
 
+    def test_create_fails_with_invalid_name(self):
+        """Test that creation fails with invalid name."""
+        c = Client()
+        zs = c.zen_store
```
[the remark this answers is not in the document]
Comment (avishniakov) [reply tests/integration/functional/zen_stores/test_zen_store.py:4106]: Used below: `c.active_user.id`

**Comment by avishniakov** [comment]
> Just nits. Otherwise looks good to me. I look to @bcdurak for the Pydantic perspective, however.

I made it in a bit more functional way, so no impact on Pydantic.

**Comment by avishniakov** [comment]
@coderabbitai review

**Review by strickvl — APPROVED** [review]

**Review by bcdurak — APPROVED** [review]
I like the changes. Thank you for making it less pydantic heavy as well, makes my job a lot easier. 😄 One question though, when it comes to named entities, I thought this approach would also go well with pipelines and pipeline runs too (not sure if I am forgetting about anything else). Is there a reason why we left those out?

Feel free to merge it. Going forward, I think we can implement a simple validator for names like this Pydantic v2 to generalize this even further.

**Comment by avishniakov** [comment]
> I like the changes. Thank you for making it less pydantic heavy as well, makes my job a lot easier. 😄 One question though, when it comes to named entities, I thought this approach would also go well with pipelines and pipeline runs too (not sure if I am forgetting about anything else). Is there a reason why we left those out?
> 
> Feel free to merge it. Going forward, I think we can implement a simple validator for names like this Pydantic v2 to generalize this even further.

Make sense, we have way more named entities indeed. I'll create a ticket with low prior to revisit this later.
The problem with making it as validator in the Pydantic is the fact that old "poorly" named entities should still work, so it might be tricky to restrict this on the model level.

**Pull request merged**